helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -0,0 +1,277 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.I0Itec.zkclient.exception.ZkMarshallingError;
+import org.I0Itec.zkclient.serialize.ZkSerializer;
+import org.apache.helix.AccessOption;
+import org.apache.helix.BaseDataAccessor;
+import org.apache.helix.BucketDataAccessor;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.client.DedicatedZkClientFactory;
+import org.apache.helix.manager.zk.client.HelixZkClient;
+import org.apache.helix.util.GZipCompressionUtil;
+
+public class ZkBucketDataAccessor implements BucketDataAccessor {
+  private static final int DEFAULT_NUM_VERSIONS = 2;
+  private static final String BUCKET_SIZE_KEY = "BUCKET_SIZE";
+  private static final String DATA_SIZE_KEY = "DATA_SIZE";
+  private static final String WRITE_LOCK_KEY = "WRITE_LOCK";
+  private static final String LAST_SUCCESS_KEY = "LAST_SUCCESS";
+
+  // 100 KB for default bucket size
+  private static final int DEFAULT_BUCKET_SIZE = 100 * 1024;
+  private final int _bucketSize;
+  private final int _numVersions;
+  private ZkSerializer _zkSerializer;
+  private BaseDataAccessor _zkBaseDataAccessor;
+  private BaseDataAccessor<ZNRecord> _znRecordBaseDataAccessor;
+
+  /**
+   * Constructor that allows a custom bucket size.
+   * @param zkAddr
+   * @param bucketSize
+   * @param numVersions number of versions to store in ZK
+   */
+  public ZkBucketDataAccessor(String zkAddr, int bucketSize, int numVersions) {
+    // There are two HelixZkClients:
+    // 1. _zkBaseDataAccessor for writes of binary data
+    // 2. _znRecordBaseDataAccessor for writes of ZNRecord (metadata)
+    HelixZkClient zkClient = DedicatedZkClientFactory.getInstance()
+        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddr));
+    zkClient.setZkSerializer(new ZkSerializer() {
+      @Override
+      public byte[] serialize(Object data) throws ZkMarshallingError {
+        if (data instanceof byte[]) {
+          return (byte[]) data;
+        }
+        throw new HelixException("ZkBucketDataAccesor only supports a byte array as an argument!");
+      }
+
+      @Override
+      public Object deserialize(byte[] data) throws ZkMarshallingError {
+        return data;
+      }
+    });
+    _zkBaseDataAccessor = new ZkBaseDataAccessor(zkClient);
+
+    // TODO: Optimize serialization with Jackson
+    // TODO: Or use a better binary serialization protocol
+    // TODO: Consider making this also binary
+    // TODO: Consider an async write for the metadata as well
+    HelixZkClient znRecordClient = DedicatedZkClientFactory.getInstance()
```

Comment:
Shall we use SharedZkClient? Note that each shared client can have separate serializer even they share the same connection. Which fits the requirement here very well.

Reply:
Changed to SharedZkClientFactory.